File: Java/UDacity/PracticeSet2/app/src/main/java/com/example/android/practiceset2/MainActivity.java
package com.example.android.practiceset2;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        String firstName = "Lyla";
        String lastName = "Fujiwara";
        String contactInfo = firstName + " " + lastName;
        contactInfo = contactInfo + "<" + lastName + "." + firstName + "@justjava.com>";
        display1(contactInfo);
    }

    void display1(String i) {
        TextView t = (TextView) findViewById(R.id.display_text_view1);
        t.setText(i);
    }

    void display2(String i) {
        TextView t = (TextView) findViewById(R.id.display_text_view2);
        t.setText(i);
    }
    void display3(String i) {
        TextView t = (TextView) findViewById(R.id.display_text_view3);
        t.setText(i);
    }
}

File: FirstLineBook/ch4/MyBestPractice/app/src/main/java/com/example/mybestpractice/MainActivity.kt
package com.example.mybestpractice

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.w3c.dom.Text

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private val msgList = ArrayList<Msg>()
    private var adapter: MsgAdapter ?= null
    lateinit var recyclerView:RecyclerView
    lateinit var inputText: TextView
    lateinit var send: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initMsg()
        val layoutManager = LinearLayoutManager(this)
        recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        inputText = findViewById<TextView>(R.id.inputText)
        send = findViewById<Button>(R.id.send)
        recyclerView.layoutManager = layoutManager
        adapter = MsgAdapter(msgList)
        recyclerView.adapter = adapter
        send.setOnClickListener(this)
    }

    private fun initMsg() {
        val msg1 = Msg("Hello guy.", Msg.TYPE_RECEIVED)
        msgList.add(msg1)
        val msg2 = Msg("Hello. Who is that?", Msg.TYPE_SENT)
        msgList.add(msg2)
        val msg3 = Msg("This is Tom. Nice talking to you. ", Msg.TYPE_RECEIVED)
        msgList.add(msg3)
    }

    override fun onClick(v: View?) {
        when(v) {
            send -> {
                val content = inputText.text.toString()
                if(content.isNotEmpty()) {
                    val msg = Msg(content, Msg.TYPE_SENT)
                    msgList.add(msg)
                    adapter?.notifyItemInserted(msgList.size-1)
                    recyclerView.scrollToPosition(msgList.size -1)
                    inputText.text = ""
                }
            }
        }
    }
}